File: 2023/08/08.exs
defmodule Day08 do
  def part1({seq, rules}) do
    seq
    |> Stream.cycle()
    |> Stream.with_index()
    |> Enum.reduce_while("AAA", fn
      {_op, i}, "ZZZ" -> {:halt, i}
      {op, _i}, acc -> {:cont, rules |> Map.get(acc) |> Enum.at(op)}
    end)
  end

  def part2({seq, rules}) do
    rules
    |> Map.keys()
    |> Enum.filter(&String.ends_with?(&1, "A"))
    |> Enum.map(fn key ->
      seq
      |> Stream.cycle()
      |> Stream.with_index()
      |> Enum.reduce_while(key, fn
        {_op, i}, <<_::binary-2, "Z">> -> {:halt, i}
        {op, _i}, acc -> {:cont, rules |> Map.get(acc) |> Enum.at(op)}
      end)
    end)
    |> lcm()
  end

  def parse(input) do
    input
    |> String.split("\n\n", trim: true)
    |> then(fn [seq, rules] ->
      {
        seq
        |> String.split("", trim: true)
        |> Enum.map(fn "L" -> 0; "R" -> 1 end),
        rules
        |> String.split("\n", trim: true)
        |> Enum.map(fn <<x::binary-3, " = (", y::binary-3, ", ", z::binary-3, ")">> -> {x,[y,z]} end)
        |> Map.new()
      }
    end)
  end

  defp gcd(a, b) when a == 0 or b == 0, do: a + b
  defp gcd(a, b), do: gcd(b, rem(a,b))
	
  defp lcm([a, b]), do: div((a*b), gcd(a,b))
  defp lcm([a | rest]), do: lcm([a, lcm(rest)])
end

input = "input" |> File.read!() |> Day08.parse()

input
|> Day08.part1()
|> IO.puts()

input
|> Day08.part2()
|> IO.puts()

File: 2023/08/08.js
const [seq, rules] = [require('fs').readFileSync('input', 'utf8').split('\n\n')].map(([seq, rules]) => [seq.split('').map(x => x == 'L' ? 0 : 1), Object.fromEntries(rules.split('\n').map(rule => rule.match(/[A-Z0-9]{3}/g)).map(([x,y,z]) => [x, [y,z]]))])[0]

let ptr = 'AAA'
for (var i = 0; ptr != 'ZZZ'; i++) ptr = rules[ptr][seq[i % seq.length]]
console.log(i)

let ptrs = Object.keys(rules).filter(x => x.endsWith('A')).map(ptr => {
  for (var i = 0; !ptr.endsWith('Z'); i++) ptr = rules[ptr][seq[i % seq.length]]
  return i
})
console.log([(f = (a,b) => b ? f(b,a%b) : a)].map(gcd => ptrs.reduce((a,b) => (b*a)/gcd(b,a)))[0])
